src/ripper/ripcddialog.cpp

```diff
@@ -241,11 +266,57 @@ void RipCDDialog::InvertSelection() {
   }
 }
 
-void RipCDDialog::Finished() { SetWorking(false); }
+void RipCDDialog::DeviceSelected(int device_index) {
+  // disconnecting from previous loader and device, if any
+  if (loader_) disconnect(loader_, nullptr, this, nullptr);
+  if (cdda_device_) disconnect(cdda_device_.get(), nullptr, this, nullptr);
+
+  ResetDialog();
+  if (device_index < 0)
+    return;  // Invalid selection, probably no devices around
+
+  Q_ASSERT(device_index < cdda_devices_.size());
+
+  DeviceInfo* device_info = cdda_devices_[device_index];
+  std::shared_ptr<ConnectedDevice> device =
+      device_manager_.Connect(device_info);
+  cdda_device_ = std::dynamic_pointer_cast<CddaDevice>(device);
+  if (!cdda_device_) {
+    rip_button_->setEnabled(false);
+    QMessageBox cdio_fail(QMessageBox::Critical, tr("Error"),
+                          tr("Failed reading CD drive"));
+    cdio_fail.exec();
+    return;
+  }
+
+  connect(cdda_device_.get(), SIGNAL(DiscChanged()), SLOT(DiscChanged()));
+
+  // get SongLoader from device and connect signals
+  loader_ = cdda_device_->loader();
+  Q_ASSERT(loader_);
+
+  connect(loader_, SIGNAL(SongsDurationLoaded(SongList)),
+          SLOT(BuildTrackListTable(SongList)));
+  connect(loader_, SIGNAL(SongsMetadataLoaded(SongList)),
+          SLOT(UpdateTrackListTable(SongList)));
+  connect(loader_, SIGNAL(SongsMetadataLoaded(SongList)),
+          SLOT(AddAlbumMetadataFromMusicBrainz(SongList)));
+
+  // load songs from new SongLoader
+  loader_->LoadSongs();
+  rip_button_->setEnabled(true);
+}
+
+void RipCDDialog::Finished() {
+  SetWorking(false);
+  Ripper* ripper = dynamic_cast<Ripper*>(sender());
```

Comment:
Good to avoid `dynamic_cast` if possible. `NewClosure` is a good way to avoid needing `sender()` at all or alternatively use a `qobject_cast` if you really have to.

Reply:
The functor variant of connect can also be used with a lambda. It feels lighter weight to me.

Reply:
done with lambda